src/Geometry/NewestVertexBisection.jl

```diff
@@ -0,0 +1,308 @@
+end
+
+"""
+node_coords == node, cell_node_ids == elem in Long Chen's notation
+"""
+function newest_vertex_bisection(
+  top::GridTopology,
+  node_coords::AbstractVector{<:VectorValue},
+  cell_node_ids::AbstractVector{<:AbstractVector{T}},
+  η_arr::AbstractArray,
+  θ::AbstractFloat,
+  sort_flag::Bool,
+) where {T <: Integer}
+  # Number of nodes
+  N::T = size(node_coords, 1)
+  elem = cell_node_ids
+  # Number of cells (triangles in 2D)
+  NT::T = size(elem, 1)
+  @assert length(η_arr) == NT
+  if sort_flag
+    _sort_longest_edge!(elem, node_coords, NT)
+  end
+  # TODO: This is rather internal so not sure how to get the testing to be
+  # outside.
+  @timeit to "_build_edges" edge = _build_edges(elem)
+  NE::T = size(edge, 1)
+  @timeit to "build_dualedge" dualedge = _build_directed_dualedge(elem, N, NT)
+  d2p = _dual_to_primal(edge, NE, N)
+  # TODO: same
+  #@timeit to "test" test_against_top(edge, top, 1)
+  @timeit to "markers" node_coords, marker =
+    _setup_markers_and_nodes!(node_coords, elem, d2p, dualedge, NE, η_arr, θ)
+  @timeit to "copy elem" elem = Vector{Vector}(elem)
+  @timeit to "bisect" cell_node_ids = _bisect(d2p, elem, marker, NT)
+  # TODO: IMPORTANT: This appears to be necessary when instatianting the RT space
+  #sort!.(cell_node_ids)
+  @timeit to "recreate" cell_node_ids = Table([c for c in cell_node_ids])
+  node_coords, cell_node_ids
+end
+
+function newest_vertex_bisection(
+  grid::Grid,
+  top::GridTopology,
+  η_arr::AbstractArray,
+  θ::AbstractFloat,
+  sort_flag::Bool,
+)
+  node_coords = get_node_coordinates(grid)
+  # TODO: Hack to make "un_lazy" version. Otherwise cannot grow dynamically
+  node_coords = [v for v in node_coords]
+  cell_node_ids = get_cell_node_ids(grid)
+  if sort_flag
+     _sort_cell_node_ids_ccw!(cell_node_ids, node_coords)
+  end
+  node_coords_ref, cell_node_ids_ref =
+    newest_vertex_bisection(top, node_coords, cell_node_ids, η_arr, θ, sort_flag)
+  reffes = get_reffes(grid)
+  cell_types = get_cell_type(grid)
+  # TODO: Same: I need to do this because I can't append to LazyVector
+  cell_types = [c for c in cell_types]
+  # TODO : Gracefully handle cell_types
+  new_cell_types = fill(1, length(cell_node_ids_ref) - length(cell_node_ids))
+  append!(cell_types, new_cell_types)
+  UnstructuredGrid(node_coords_ref, cell_node_ids_ref, reffes, cell_types)
+end
+
+"""
+The newest vertex bisection algorithm provides a method of local refinement
+without creating hanging nodes. For now, only 2D simplicial meshes are
+supported.
+
+This implementation is based on the following article:
+
+LONG CHEN (2008). SHORT IMPLEMENTATION OF BISECTION IN MATLAB.
+In Recent Advances in Computational Sciences. WORLD SCIENTIFIC.
+DOI: 10.1142/9789812792389_0020
+
+# Arguments
+
+ -`model::DiscreteModel`: The current DiscreteModel to be refined.
+
+ -`η_arr::AbstractArray`: The values of the estimator on each cell of the domain
+   ordered in a consistent way with Geometry.Gridap.get_faces.
```

Comment:
Which is the length of this array? num_cells(model) ? If yes, mention this instead of the relation with get_faces.